test/dynamo/test_optimizers.py

```diff
@@ -95,3 +95,37 @@ def setUpClass(cls):
 
 for opt in optimizers:
     setattr(OptimizerTests, "test_" + opt.__name__.lower(), make_test(opt))
+
+
+class End2EndTests(torchdynamo.testing.TestCase):
+
+    # https://github.com/pytorch/torchdynamo/issues/1604
+    def test_optimizing_over_tensor_with_requires_grad(self):
+        class Net(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                z = torch.bmm(x, y)
+                z = torch.flatten(z, 1)
+                return z
+
+        def training_iter_fn(batch, model, optimizer):
+            optimizer.zero_grad()
+            out = model(**batch)
+            target = torch.tensor([0, 7])
+            loss = torch.nn.CrossEntropyLoss()(out, target)
+            loss.backward()
+            optimizer.step()
+            return loss
+
+        net = Net()
+        input1 = torch.randn(2, 1, 4)
+        input2 = torch.randn(2, 4, 8, requires_grad=True)
+        optimizer = torch.optim.Adam([input2], lr=0.1)
+
+        cnts = torchdynamo.testing.CompileCounter()
+        opt_training_iter_fn = torchdynamo.optimize(cnts)(training_iter_fn)
+        batch = {"x": input1, "y": input2}
+        opt_training_iter_fn(batch, net, optimizer)
+        self.assertEqual(cnts.frame_count, 4)
```

Comment:
Why is this 4 frames? Where are the breaks?

Reply:
Yea, I think these breaks are known issues:
```
[2022-10-14 05:19:22,296] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.backward from user code at   File "/scratch/ybliang/work/repos/torchdynamo/debug/debug1.py", line 53, in training_iter_fn
    loss.backward()

[2022-10-14 05:19:22,381] torchdynamo.symbolic_convert: [WARNING] Graph break: inline with __closure__ from user code at   File "/scratch/ybliang/work/repos/torchdynamo/debug/debug1.py", line 54, in <graph break in training_iter_fn>
    optimizer.step()

[2022-10-14 05:19:22,392] torchdynamo.symbolic_convert: [WARNING] Graph break: inline in skipfiles: _fn /scratch/ybliang/work/repos/torchdynamo/torchdynamo/eval_frame.py from user code at   File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 178, in step
    self._cuda_graph_capture_health_check()

[2022-10-14 05:19:22,428] torchdynamo.symbolic_convert: [WARNING] Graph break: Tensor.item from user code at   File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 234, in <graph break in step>
    adam(params_with_grad,
  File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 300, in adam
    func(params,
  File "/scratch/ybliang/work/env/lib/python3.9/site-packages/torch/optim/adam.py", line 395, in _single_tensor_adam
    step = step_t.item()
```
Some of the breaks, e.g, ```Tensor.item``` actually we can avoid by setting ```capturable=True```, but here I use the basic API for unit test.